src/main/groovy/com/calgaryscientific/gradle/VeracodeTask.groovy

```diff
@@ -44,50 +44,45 @@ abstract class VeracodeTask extends DefaultTask {
             'fileId'           : '123',
             'mode'             : 'action|actionSummary|verbose',
             'maxUploadAttempts': '123',
-            'fileId'           : 'xxx',
             'buildId1'         : '123',
             'buildId2'         : '123'
     ]
 
-    def requiredArguments = []
+    List<String> requiredArguments = []
+    List<String> optionalArguments = []
 
     VeracodeTask() {
         group = 'Veracode'
     }
 
     abstract void run()
 
-    final String correctUsage() {
-        StringBuilder sb = new StringBuilder("Example of usage: gradle ${getName()}")
-        requiredArguments.each() {
-            if (!isArgumentOptional(it)) {
-                sb.append(" -P${it}=${validArguments.get(it)}")
-            } else {
-                String originalArgument = it.substring(0, it.length() - OPTIONAL.length())
-                sb.append(" [-P${originalArgument}=${validArguments.get(originalArgument)}]")
-            }
+    protected static String correctUsage(String taskName,
+                                         List<String> requiredArguments,
+                                         List<String> optionalArguments) {
+        StringBuilder sb = new StringBuilder("Missing required arguments: gradle ${taskName}")
+        requiredArguments.each() { arg ->
+            sb.append(" -P${arg}=${validArguments.get(arg)}")
+        }
+        optionalArguments.each() { arg ->
+            sb.append(" [-P${arg}=${validArguments.get(arg)}]")
         }
-
         sb.toString()
     }
 
-    final boolean haveRequiredArguments() {
-        boolean haveRequiredArguments = true
-        requiredArguments.each() {
-            if (!isArgumentOptional(it)) {
-                haveRequiredArguments &= getProject().hasProperty(it)
-            }
+    protected boolean hasRequiredArguments() {
+        boolean hasRequiredArguments = true
+        requiredArguments.each() { arg ->
+            hasRequiredArguments &= getProject().hasProperty(arg)
```

Comment:
optional, you can try to use groovy idiom here: requiredArguments.any { arg -> !getProject).hasProperty(arg) }